lib/promscrape/scrapework.go

```diff
@@ -745,42 +776,51 @@ func (sw *scrapeWork) sendStaleSeries(lastScrape, currScrape string, timestamp i
 	if currScrape != "" {
 		bodyString = parser.GetRowsDiff(lastScrape, currScrape)
 	}
-	wc := writeRequestCtxPool.Get(sw.prevLabelsLen)
-	defer func() {
-		wc.reset()
-		writeRequestCtxPool.Put(wc)
-	}()
 	if bodyString != "" {
 		// Send stale markers in streaming mode in order to reduce memory usage
 		// when stale markers for targets exposing big number of metrics must be generated.
 		// See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3668
 		// and https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3675
-		var mu sync.Mutex
 		br := bytes.NewBufferString(bodyString)
 		err := stream.Parse(br, timestamp, "", false, func(rows []parser.Row) error {
-			mu.Lock()
-			defer mu.Unlock()
+			// stream.Parse internally allocates at least a 64KB buffer.
+			// To minimize memory allocations and improve throughput, we preallocate
+			// a buffer of the same size.
+			//
+			// See https://github.com/VictoriaMetrics/VictoriaMetrics/blob/f645479b5e0612e7e781aa50a0375c3624b3adfe/lib/protoparser/protoparserutil/lines_reader.go#L42
+			wc := writeRequestCtxPool.Get(protoparserutil.DefaultBlockSize)
+			defer func() {
+				wc.resetNoRows()
+				writeRequestCtxPool.Put(wc)
+			}()
+
 			for i := range rows {
 				sw.addRowToTimeseries(wc, &rows[i], timestamp, true)
 			}
+
 			// Apply series limit to stale markers in order to prevent sending stale markers for newly created series.
 			// See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3660
-			if sw.seriesLimitExceeded {
+			if sw.seriesLimitExceeded.Load() {
```

Comment:
is seriesLimitExceeded is modified only in `applySeriesLimit`, maybe it should be checked there as well?

Reply:
do you mean the condition should be moved inside applySeriesLimit? If so, it would not work, because there are other conditions for calling the function, like https://github.com/VictoriaMetrics/VictoriaMetrics/blob/dec237b7d695f83c1a4b716e5e8c167dd1c670fb/lib/promscrape/scrapework.go#L504

Reply:
the `!areIdenticalSeries` can still be checked by the caller.

Reply:
Sorry, I don’t see how we can split this `if` condition. The check ensures that we call applySeriesLimit when either seriesLimitExceeded is true or !areIdenticalSeries. In other words, it is an  `OR` condition. Moving one part inside the function would change the behavior to `AND` condition, as the caller would no longer control when applySeriesLimit is invoked under either condition. Let me know if I’m missing something!

IMO I prefer the control to be outside the function, it helps the function's reusability.

Reply:
Thanks for explanation. I do agree.